File: go/webhook.go
package plutus

import (
	"bytes"
	"crypto/rsa"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strconv"
	"time"
)

// MaxWebhookBodyBytes 是读取 Webhook 请求体的硬上限, 防止无界读取。
const MaxWebhookBodyBytes = 4 << 20

// WebhookBodyDigestBase64 计算 Webhook 签名规范串第 4 行的 body 摘要。
// 注意是 Base64, 与 API 链的小写 hex 摘要不是同一套编码。
func WebhookBodyDigestBase64(body []byte) string {
	sum := sha256.Sum256(body)
	return base64.StdEncoding.EncodeToString(sum[:])
}

// WebhookHeaders 是 Webhook 投递的传输头。
type WebhookHeaders struct {
	// DeliveryID 取自 X-SlaunchX-Delivery-Id, 是去重键。
	DeliveryID string
	// EventType 取自 X-SlaunchX-Event-Type。
	EventType string
	// Timestamp 取自 X-SlaunchX-Timestamp, Unix 毫秒。
	Timestamp string
	// KeyID 取自 X-SlaunchX-Key-Id, 是接收方 API Key 业务 ID, 不是指纹。
	KeyID string
	// Signature 取自 X-SlaunchX-Signature, Base64 RSA-SHA256 签名。
	Signature string
}

// WebhookHeadersFromHTTP 从 http.Header 提取 Webhook 传输头。
func WebhookHeadersFromHTTP(h http.Header) WebhookHeaders {
	return WebhookHeaders{
		DeliveryID: h.Get(HeaderWebhookDeliveryID),
		EventType:  h.Get(HeaderWebhookEventType),
		Timestamp:  h.Get(HeaderWebhookTimestamp),
		KeyID:      h.Get(HeaderWebhookKeyID),
		Signature:  h.Get(HeaderWebhookSignature),
	}
}

// CanonicalString 返回 Webhook 签名规范串 (4 行)。
func (h WebhookHeaders) CanonicalString(body []byte) string {
	return WebhookCanonicalString(h.DeliveryID, h.EventType, h.Timestamp, WebhookBodyDigestBase64(body))
}

// AAD 返回 Webhook 信封的 AAD 四元组: routeTemplate 固定为字面量 "webhook",
// keyId 为 API Key 业务 ID。
func (h WebhookHeaders) AAD() AAD {
	return AAD{
		RequestID:     h.DeliveryID,
		RouteTemplate: WebhookRouteTemplate,
		Timestamp:     h.Timestamp,
		KeyID:         h.KeyID,
	}
}

// WebhookPayload 是解密后的通知明文的顶层结构。
//
// Data 保持为原始 JSON, 由调用方按事件类型自行解码。金额是 {currency, amount} 且
// amount 为十进制字符串, 不得解析为浮点数, 见 Amount。
type WebhookPayload struct {
	EventID              string          `json:"eventId"`
	EventType            string          `json:"eventType"`
	PayloadSchemaVersion int             `json:"payloadSchemaVersion"`
	OccurredAt           string          `json:"occurredAt"`
	WorkspaceBizID       string          `json:"workspaceBizId"`
	DeliveryBizID        string          `json:"deliveryBizId"`
	Resource             WebhookResource `json:"resource"`
	Data                 json.RawMessage `json:"data"`
}

// WebhookResource 是通知指向的资源标识。
type WebhookResource struct {
	Type  string `json:"type"`
	BizID string `json:"bizId"`
}

// Amount 是协议中的金额表示。Amount 字段是主单位十进制字符串, 不是浮点数。
type Amount struct {
	Currency string `json:"currency"`
	Amount   string `json:"amount"`
}

// WebhookNotification 是一次通过验签与解密的 Webhook 投递。
type WebhookNotification struct {
	// Headers 是传输头。
	Headers WebhookHeaders
	// Body 是原始 HTTP body (即签名覆盖的信封 JSON 字节)。
	Body []byte
	// Envelope 是解析后的 Webhook 信封。
	Envelope *Envelope
	// Plaintext 是解密后的通知明文字节。
	Plaintext []byte
	// Payload 是解析后的通知载荷。
	Payload WebhookPayload
}

// DecodeData 把通知的 data 字段解码到 v。
func (n *WebhookNotification) DecodeData(v any) error {
	if len(n.Payload.Data) == 0 {
		return fmt.Errorf("plutus: webhook payload has no data field")
	}
	return json.Unmarshal(n.Payload.Data, v)
}

// ParseWebhookEnvelope 严格解析 Webhook 信封 JSON。
//
// Webhook 信封的 schema 为 additionalProperties:false, 因此出现未知字段 (含 encryptedPayload)
// 即视为非法; envelopeVersion 必须恰为 1。
func ParseWebhookEnvelope(body []byte) (*Envelope, error) {
	var raw struct {
		EnvelopeVersion *int   `json:"envelopeVersion"`
		Algorithm       string `json:"algorithm"`
		KeyFingerprint  string `json:"keyFingerprint"`
		EncryptedKey    string `json:"encryptedKey"`
		Ciphertext      string `json:"ciphertext"`
		AAD             string `json:"aad"`
	}
	dec := json.NewDecoder(bytes.NewReader(body))
	dec.DisallowUnknownFields()
	if err := dec.Decode(&raw); err != nil {
		return nil, fmt.Errorf("%w: %v", ErrEnvelopeInvalid, err)
	}
	if raw.EnvelopeVersion == nil || *raw.EnvelopeVersion != 1 {
		return nil, fmt.Errorf("%w: envelopeVersion must be 1", ErrEnvelopeInvalid)
	}
	env := &Envelope{
		EnvelopeVersion: raw.EnvelopeVersion,
		Algorithm:       raw.Algorithm,
		KeyFingerprint:  raw.KeyFingerprint,
		EncryptedKey:    raw.EncryptedKey,
		Ciphertext:      raw.Ciphertext,
		AAD:             raw.AAD,
	}
	if err := env.Validate(); err != nil {
		return nil, err
	}
	return env, nil
}

// WebhookConfig 是 WebhookReceiver 的配置。密钥可用 PEM 或已解析对象提供, 两者取其一。
type WebhookConfig struct {
	// APIKey 是本商户的 API Key 业务 ID, 即 AAD 第 4 分量的期望值。
	// 非空时会与 X-SlaunchX-Key-Id 比对, 不一致直接拒绝。
	APIKey string
	// PlatformAuthPublicKeyPEM 是平台认证公钥, 用于验签。
	PlatformAuthPublicKeyPEM []byte
	// PlatformAuthPublicKey 是已解析的平台认证公钥。
	PlatformAuthPublicKey *rsa.PublicKey
	// MerchantEncPrivateKeyPEM 是商户加密私钥, 用于解密。
	MerchantEncPrivateKeyPEM []byte
	// MerchantEncPrivateKey 是已解析的商户加密私钥。
	MerchantEncPrivateKey *rsa.PrivateKey
	// TimestampTolerance 是本地时间戳容差; 0 表示不校验。
	// 协议未规定 Webhook 时间戳窗口 (SPEC 第 16 节), 默认关闭。
	TimestampTolerance time.Duration
	// Now 用于测试注入当前时间; 默认 time.Now。
	Now func() time.Time
}

// WebhookReceiver 完成 Webhook 的验签、解密与交叉校验。可并发使用。
type WebhookReceiver struct {
	apiKey      string
	verifyKey   *rsa.PublicKey
	decryptKey  *rsa.PrivateKey
	fingerprint string
	tolerance   time.Duration
	now         func() time.Time
}

// NewWebhookReceiver 构造 Webhook 接收器。
func NewWebhookReceiver(cfg WebhookConfig) (*WebhookReceiver, error) {
	verifyKey := cfg.PlatformAuthPublicKey
	if verifyKey == nil {
		if len(cfg.PlatformAuthPublicKeyPEM) == 0 {
			return nil, fmt.Errorf("%w: platform_auth public key is required", ErrInvalidConfig)
		}
		parsed, err := ParseRSAPublicKeyPEM(cfg.PlatformAuthPublicKeyPEM)
		if err != nil {
			return nil, err
		}
		verifyKey = parsed
	}
	decryptKey := cfg.MerchantEncPrivateKey
	if decryptKey == nil {
		if len(cfg.MerchantEncPrivateKeyPEM) == 0 {
			return nil, fmt.Errorf("%w: merchant_enc private key is required", ErrInvalidConfig)
		}
		parsed, err := ParseRSAPrivateKeyPEM(cfg.MerchantEncPrivateKeyPEM)
		if err != nil {
			return nil, err
		}
		decryptKey = parsed
	}
	fingerprint, err := KeyFingerprint(&decryptKey.PublicKey)
	if err != nil {
		return nil, err
	}
	now := cfg.Now
	if now == nil {
		now = time.Now
	}
	return &WebhookReceiver{
		apiKey:      cfg.APIKey,
		verifyKey:   verifyKey,
		decryptKey:  decryptKey,
		fingerprint: fingerprint,
		tolerance:   cfg.TimestampTolerance,
		now:         now,
	}, nil
}

// Verify 只做验签: 重建 4 行规范串并用平台认证公钥验证 X-SlaunchX-Signature。
// 签名覆盖的是信封 JSON 的原始字节, 必须先验签再解密。
func (r *WebhookReceiver) Verify(headers WebhookHeaders, body []byte) error {
	if headers.Signature == "" {
		return fmt.Errorf("%w: signature header is missing", ErrWebhookSignatureInvalid)
	}
	canonical := headers.CanonicalString(body)
	if err := VerifyCanonicalString(r.verifyKey, canonical, headers.Signature); err != nil {
		return fmt.Errorf("%w: %v", ErrWebhookSignatureInvalid, err)
	}
	return nil
}

// Handle 完成一次投递的全部处理: 验签 → 解析信封 → 重建 AAD 解密 → 解析明文 → 交叉校验。
//
// 调用方仍需按 Payload.DeliveryBizID 做去重, 自动重试与人工重放共用同一个投递 ID。
func (r *WebhookReceiver) Handle(headers WebhookHeaders, body []byte) (*WebhookNotification, error) {
	if r.apiKey != "" && headers.KeyID != r.apiKey {
		return nil, fmt.Errorf("%w: %s is %q, expected %q", ErrWebhookPayloadMismatch, HeaderWebhookKeyID, headers.KeyID, r.apiKey)
	}
	if err := r.checkTimestamp(headers.Timestamp); err != nil {
		return nil, err
	}
	if err := r.Verify(headers, body); err != nil {
		return nil, err
	}
	env, err := ParseWebhookEnvelope(body)
	if err != nil {
		return nil, err
	}
	if err := env.VerifyFingerprint(r.fingerprint); err != nil {
		return nil, err
	}
	plaintext, err := Open(r.decryptKey, env, headers.AAD())
	if err != nil {
		return nil, err
	}
	var payload WebhookPayload
	if err := json.Unmarshal(plaintext, &payload); err != nil {
		return nil, fmt.Errorf("plutus: decode webhook payload: %w", err)
	}
	if payload.DeliveryBizID != headers.DeliveryID {
		return nil, fmt.Errorf("%w: deliveryBizId %q != header %q", ErrWebhookPayloadMismatch, payload.DeliveryBizID, headers.DeliveryID)
	}
	if payload.EventType != headers.EventType {
		return nil, fmt.Errorf("%w: eventType %q != header %q", ErrWebhookPayloadMismatch, payload.EventType, headers.EventType)
	}
	if payload.PayloadSchemaVersion != 1 {
		return nil, fmt.Errorf("%w: payloadSchemaVersion must be 1, got %d", ErrWebhookPayloadMismatch, payload.PayloadSchemaVersion)
	}
	return &WebhookNotification{
		Headers:   headers,
		Body:      body,
		Envelope:  env,
		Plaintext: plaintext,
		Payload:   payload,
	}, nil
}

// HandleRequest 从 http.Request 读取原始 body 并处理。
//
// 必须在任何中间件解析 body 之前调用: 签名覆盖的是原始字节, 重新序列化会破坏摘要。
func (r *WebhookReceiver) HandleRequest(req *http.Request) (*WebhookNotification, error) {
	if req.Body == nil {
		return nil, fmt.Errorf("%w: request body is nil", ErrWebhookSignatureInvalid)
	}
	body, err := io.ReadAll(io.LimitReader(req.Body, MaxWebhookBodyBytes+1))
	if err != nil {
		return nil, fmt.Errorf("plutus: read webhook body: %w", err)
	}
	if len(body) > MaxWebhookBodyBytes {
		return nil, fmt.Errorf("plutus: webhook body exceeds %d bytes", MaxWebhookBodyBytes)
	}
	return r.Handle(WebhookHeadersFromHTTP(req.Header), body)
}

func (r *WebhookReceiver) checkTimestamp(timestamp string) error {
	if r.tolerance <= 0 {
		return nil
	}
	millis, err := strconv.ParseInt(timestamp, 10, 64)
	if err != nil {
		return fmt.Errorf("%w: %s is not a decimal millisecond string", ErrInvalidTimestamp, HeaderWebhookTimestamp)
	}
	delta := r.now().Sub(time.UnixMilli(millis))
	if delta < 0 {
		delta = -delta
	}
	if delta > r.tolerance {
		return fmt.Errorf("%w: skew %s exceeds %s", ErrWebhookTimestampOutOfRange, delta, r.tolerance)
	}
	return nil
}

File: node/src/webhook.ts
/**
 * Webhook 接收:验签(Base64 body 摘要)+ 信封解密(AAD 第二分量固定 `webhook`)。
 *
 * 顺序不可颠倒:签名覆盖的是**加密信封 JSON 的原始字节**,必须先验签、后解密。
 */

import { timingSafeEqual, type KeyObject } from 'node:crypto';
import { buildWebhookCanonicalString, webhookBodyDigestBase64 } from './canonical.js';
import { EnvelopeCodec, ENVELOPE_ALGORITHM, type WebhookEnvelope } from './envelope.js';
import { PlutusWebhookError } from './errors.js';
import { keyFingerprint, loadPrivateKey, loadPublicKey, type KeyInput } from './keys.js';
import { verifyCanonicalSignature } from './signer.js';
import { readHeader, type HeaderSource } from './verifier.js';

/** AAD 第二分量的字面量。 */
export const WEBHOOK_ROUTE_TEMPLATE = 'webhook';

/** Webhook 传输头。 */
export interface WebhookHeaders {
  /** `X-SlaunchX-Delivery-Id`,投递 ID,重试与重放保持不变,是去重键 */
  deliveryBizId: string;
  /** `X-SlaunchX-Event-Type` */
  eventType: string;
  /** `X-SlaunchX-Timestamp`,Unix 毫秒 */
  timestamp: string;
  /** `X-SlaunchX-Key-Id`,接收方 API Key 业务 ID(不是指纹) */
  keyId: string;
  /** `X-SlaunchX-Signature`,Base64 */
  signature: string;
}

/** 明文通知载荷的顶层结构(SPEC 10.5)。 */
export interface WebhookPayload {
  eventId: string;
  eventType: string;
  payloadSchemaVersion: number;
  /** UTC RFC 3339,带 `Z` */
  occurredAt: string;
  workspaceBizId: string;
  deliveryBizId: string;
  resource: { type: string; bizId: string } & Record<string, unknown>;
  data: Record<string, unknown>;
  [key: string]: unknown;
}

/** 完整的 Webhook 处理结果。 */
export interface WebhookDelivery<T = WebhookPayload> {
  /** 已校验的传输头 */
  headers: WebhookHeaders;
  /** 解析后的信封 */
  envelope: WebhookEnvelope;
  /** 解密后的明文通知载荷 */
  payload: T;
  /** 明文原文,便于原样落库 */
  plaintext: string;
}

/** {@link WebhookHandler} 构造选项。 */
export interface WebhookHandlerOptions {
  /** 平台认证公钥(`platform_auth`),用于验签 */
  platformAuthPublicKey: KeyInput;
  /** 商户加密私钥(`merchant_enc`),用于解密 */
  merchantEncPrivateKey: KeyInput;
  /**
   * 本商户的 API Key 业务 ID，必填。强制校验 `X-SlaunchX-Key-Id` 与之相等,
   * 并用于构造 AAD 的第四分量。
   */
  apiKeyBizId: string;
  /**
   * 允许的时间戳偏差(毫秒)。SPEC 第 16 节明确该窗口未在契约中规定,
   * 因此**默认关闭**(`undefined`)。启用后既拒绝过旧也拒绝过新的投递。
   */
  timestampToleranceMs?: number;
  /** 毫秒时钟,默认 `Date.now` */
  now?: () => number;
  /** 是否交叉校验明文与传输头(SPEC 10.6),默认 `true` */
  crossCheckPayload?: boolean;
  /** 明文长度上限,默认 1 MiB */
  maxPlaintextBytes?: number;
  /** 是否执行严格密钥格式校验,默认 `true` */
  strictKeyValidation?: boolean;
}

function requireHeader(headers: HeaderSource, name: string): string {
  const value = readHeader(headers, name);
  if (!value) {
    throw new PlutusWebhookError(`missing webhook header ${name}`, 'SDK.WEBHOOK_HEADER_MISSING');
  }
  return value;
}

/**
 * 从传输头中提取并校验 Webhook 头集合。
 *
 * @param headers - HTTP 头
 * @returns 五个必需头的取值
 * @throws {PlutusWebhookError} 缺少必需头
 */
export function extractWebhookHeaders(headers: HeaderSource): WebhookHeaders {
  return {
    deliveryBizId: requireHeader(headers, 'X-SlaunchX-Delivery-Id'),
    eventType: requireHeader(headers, 'X-SlaunchX-Event-Type'),
    timestamp: requireHeader(headers, 'X-SlaunchX-Timestamp'),
    keyId: requireHeader(headers, 'X-SlaunchX-Key-Id'),
    signature: requireHeader(headers, 'X-SlaunchX-Signature'),
  };
}

/**
 * 校验 Webhook 信封形状(SPEC 10.3)。
 *
 * 必需字段:`envelopeVersion`(恰为 1)、`algorithm`、`keyFingerprint`、`encryptedKey`、
 * `ciphertext`、`aad`;schema 声明 `additionalProperties: false`,因此出现
 * `encryptedPayload` 等额外字段一律拒绝。
 *
 * @param value - 已解析的 body JSON
 * @returns 通过校验的信封
 * @throws {PlutusWebhookError} 形状不符
 */
export function assertWebhookEnvelope(value: unknown): WebhookEnvelope {
  if (!value || typeof value !== 'object' || Array.isArray(value)) {
    throw new PlutusWebhookError('webhook body must be a JSON object');
  }
  const record = value as Record<string, unknown>;
  const allowed = new Set(['envelopeVersion', 'algorithm', 'keyFingerprint', 'encryptedKey', 'ciphertext', 'aad']);
  for (const key of Object.keys(record)) {
    if (!allowed.has(key)) {
      throw new PlutusWebhookError(`webhook envelope contains an unexpected field: ${key}`);
    }
  }
  if (record['envelopeVersion'] !== 1) {
    throw new PlutusWebhookError(`webhook envelopeVersion must be exactly 1, got ${String(record['envelopeVersion'])}`);
  }
  if (record['algorithm'] !== ENVELOPE_ALGORITHM) {
    throw new PlutusWebhookError(`webhook envelope algorithm must be ${ENVELOPE_ALGORITHM}`);
  }
  for (const key of ['keyFingerprint', 'encryptedKey', 'ciphertext', 'aad'] as const) {
    if (typeof record[key] !== 'string' || (record[key] as string).length === 0) {
      throw new PlutusWebhookError(`webhook envelope field "${key}" must be a non-empty string`);
    }
  }
  return record as unknown as WebhookEnvelope;
}

/**
 * Webhook 处理器。
 *
 * @example
 * const handler = new WebhookHandler({ platformAuthPublicKey, merchantEncPrivateKey, apiKeyBizId });
 * // Express: 必须拿到未被解析器改写的 raw body
 * app.post('/webhooks/slaunchx', express.raw({ type: 'application/json' }), (req, res) => {
 *   const delivery = handler.handle(req.body, req.headers);
 *   res.status(204).end();
 * });
 */
export class WebhookHandler {
  private readonly publicKey: KeyObject;
  private readonly privateKey: KeyObject;
  private readonly merchantEncFingerprint: string;
  private readonly apiKeyBizId: string;
  private readonly timestampToleranceMs: number | undefined;
  private readonly now: () => number;
  private readonly crossCheckPayload: boolean;
  private readonly maxPlaintextBytes: number | undefined;

  constructor(options: WebhookHandlerOptions) {
    if (typeof options.apiKeyBizId !== 'string' || !options.apiKeyBizId.trim()) {
      throw new PlutusWebhookError('apiKeyBizId is required');
    }
    const strict = options.strictKeyValidation !== false;
    this.publicKey = loadPublicKey(options.platformAuthPublicKey, { strict, label: 'platform_auth' });
    this.privateKey = loadPrivateKey(options.merchantEncPrivateKey, { strict, label: 'merchant_enc' });
    this.merchantEncFingerprint = keyFingerprint(this.privateKey);
    this.apiKeyBizId = options.apiKeyBizId;
    this.timestampToleranceMs = options.timestampToleranceMs;
    this.now = options.now ?? (() => Date.now());
    this.crossCheckPayload = options.crossCheckPayload !== false;
    this.maxPlaintextBytes = options.maxPlaintextBytes;
  }

  /** 商户加密公钥指纹,应与信封 `keyFingerprint` 相等。 */
  get keyFingerprint(): string {
    return this.merchantEncFingerprint;
  }

  /**
   * 只验签,不解密。
   *
   * @param rawBody - **未经任何解析器改写**的原始 HTTP body 字节
   * @param headers - HTTP 头
   * @returns 已校验的传输头
   * @throws {PlutusWebhookError} 头缺失、Key-Id 不匹配、时间戳超窗、验签失败
   */
  verify(rawBody: Uint8Array, headers: HeaderSource): WebhookHeaders {
    const parsed = extractWebhookHeaders(headers);
    if (this.timestampToleranceMs !== undefined) {
      if (!/^\d+$/.test(parsed.timestamp)) {
        throw new PlutusWebhookError('X-SlaunchX-Timestamp must be a Unix milliseconds decimal string');
      }
      const skew = Math.abs(this.now() - Number(parsed.timestamp));
      if (skew > this.timestampToleranceMs) {
        throw new PlutusWebhookError(
          `webhook timestamp skew ${skew}ms exceeds the configured tolerance ${this.timestampToleranceMs}ms`,
          'SDK.WEBHOOK_TIMESTAMP_SKEW',
        );
      }
    }
    const canonical = buildWebhookCanonicalString(
      parsed.deliveryBizId,
      parsed.eventType,
      parsed.timestamp,
      webhookBodyDigestBase64(rawBody),
    );
    if (!verifyCanonicalSignature(canonical, parsed.signature, this.publicKey)) {
      throw new PlutusWebhookError('webhook signature verification failed', 'SDK.WEBHOOK_SIGNATURE_INVALID');
    }
    const expected = Buffer.from(this.apiKeyBizId, 'utf8');
    const received = Buffer.from(parsed.keyId, 'utf8');
    if (expected.length !== received.length || !timingSafeEqual(expected, received)) {
      throw new PlutusWebhookError('X-SlaunchX-Key-Id does not match the configured API key business id');
    }
    return parsed;
  }

  /**
   * 完整处理:验签 → 解析信封 → 重建 AAD → 解密 → 交叉校验明文。
   *
   * 调用方仍需按 `deliveryBizId` 去重(自动重试与人工重放共用同一个 `deliveryBizId`)。
   *
   * @param rawBody - 原始 HTTP body 字节
   * @param headers - HTTP 头
   * @returns 头、信封、明文载荷
   */
  handle<T = WebhookPayload>(rawBody: Uint8Array, headers: HeaderSource): WebhookDelivery<T> {
    const parsedHeaders = this.verify(rawBody, headers);

    let json: unknown;
    try {
      json = JSON.parse(Buffer.from(rawBody).toString('utf8'));
    } catch (cause) {
      throw new PlutusWebhookError('webhook body is not valid JSON', 'SDK.WEBHOOK_INVALID', { cause });
    }
    const envelope = assertWebhookEnvelope(json);
    if (envelope.keyFingerprint !== this.merchantEncFingerprint) {
      throw new PlutusWebhookError(
        'webhook envelope keyFingerprint does not match the configured merchant_enc key',
      );
    }

    const plaintext = EnvelopeCodec.open(envelope, {
      privateKey: this.privateKey,
      aad: {
        requestId: parsedHeaders.deliveryBizId,
        routeTemplate: WEBHOOK_ROUTE_TEMPLATE,
        timestamp: parsedHeaders.timestamp,
        keyId: this.apiKeyBizId,
      },
      expectedKeyFingerprint: this.merchantEncFingerprint,
      maxPlaintextBytes: this.maxPlaintextBytes,
    }).toString('utf8');

    let payload: T;
    try {
      payload = JSON.parse(plaintext) as T;
    } catch (cause) {
      throw new PlutusWebhookError('decrypted webhook payload is not valid JSON', 'SDK.WEBHOOK_INVALID', { cause });
    }
    if (this.crossCheckPayload) {
      assertPayloadMatchesHeaders(payload as unknown as Partial<WebhookPayload>, parsedHeaders);
    }
    return { headers: parsedHeaders, envelope, payload, plaintext };
  }
}

/**
 * 交叉校验明文与传输头(SPEC 10.6)。
 *
 * @param payload - 解密后的明文载荷
 * @param headers - 传输头
 * @throws {PlutusWebhookError} `deliveryBizId` / `eventType` 不一致,或 `payloadSchemaVersion` 不为 1
 */
export function assertPayloadMatchesHeaders(
  payload: Partial<WebhookPayload>,
  headers: WebhookHeaders,
): void {
  if (payload.deliveryBizId !== headers.deliveryBizId) {
    throw new PlutusWebhookError('payload deliveryBizId does not match X-SlaunchX-Delivery-Id');
  }
  if (payload.eventType !== headers.eventType) {
    throw new PlutusWebhookError('payload eventType does not match X-SlaunchX-Event-Type');
  }
  if (payload.payloadSchemaVersion !== 1) {
    throw new PlutusWebhookError(
      `unsupported payloadSchemaVersion: ${String(payload.payloadSchemaVersion)}`,
      'SDK.WEBHOOK_SCHEMA_UNSUPPORTED',
    );
  }
}
